### PR TITLE
Allow amending the last commit

### LINKED ISSUE #1644: Enhancement: Amend last commit
### Description

Amend last commit feature is very useful and exists in many git clients, but not in GitHub Desktop.
It would be nice if we have it in GitHub Desktop too.

Current workaround is click `Undo` and commit again but I've to carefully check what I want to commit again.

### Version
**GitHub Desktop version:** 0.5.5

Thanks.

### PR DESCRIPTION
**Pull request by sergiou87**
Closes #1644 

## Description

This PR introduces the ability to amend the last commit in your branch (if it's not pushed) by right-clicking it in the `History` tab and selecting the new `Amend Commit…` action.

This will switch Desktop to a "amending" state, where all selected changes and the new summary and description will be applied to the last commit.

### Screenshots

https://user-images.githubusercontent.com/1083228/120630029-b2c69880-c466-11eb-9c48-dafc769459d0.mov

## Release notes

Notes: [New] Amend the last local commit in your branch

### CODE REVIEW

File: app/src/lib/stores/repository-state-cache.ts [hunk 1]
```diff
@@ -46,7 +46,24 @@ export class RepositoryStateCache {
...
+    // and there is no "fixing conflicts" state.
+    const newIsAmending =
+      newState.isAmending &&
+      isSameLastLocalCommit &&
+      newState.changesState.conflictState === null
+
+    this.repositoryState.set(repository.hash, {
+      ...newState,
+      isAmending: newIsAmending,
+    })
```
Comment (sergiou87) [inline app/src/lib/stores/repository-state-cache.ts:66]: If you check [my previous approach](https://github.com/desktop/desktop/pull/12353/commits/b902f8e146ada6182d7bf6e6c07034500f75a5ee), I intended to stop amending manually whenever the user started any operation that could lead to a state where the user would amend the wrong commit (for example, you switch to a completely different branch… the amend state should be cleared… and same for pulling, pushing, merging, cherry-picking, squashing…).

After thinking a bit about it, I decided to go with this other centralized and (more) future-proof approach, where the amending state will be cleared whenever:
1. The last local commit sha changes
2. The app is in a "fix conflicts" state

This should cover all our present needs, at least!

Comment (tidy-dev) [reply app/src/lib/stores/repository-state-cache.ts:66]: Nice!

File: app/src/ui/changes/commit-message.tsx [hunk 2]
```diff
@@ -628,17 +673,26 @@ export class CommitMessage extends React.Component<
...
-      )
+    const defaultContents = isAmending ? (
+      <>{commitVerb} last commit</>
+    ) : branchName !== null ? (
+      <>
+        {commitVerb} to <strong>{branchName}</strong>
+      </>
+    ) : (
+      commitVerb
+    )
```
Comment (tidy-dev) [inline app/src/ui/changes/commit-message.tsx]: Suggestion. I found this hard to read.. maybe worth separating into two variables. commitDefaultContents and amendDefaultContents and then defaultContents = isAmending ? amendDefaultContents : commitDefaultContents

File: app/src/ui/changes/commit-message.tsx [hunk 3]
```diff
@@ -628,17 +673,26 @@ export class CommitMessage extends React.Component<
     const { isCommitting, branch: branchName, commitButtonText } = this.props
 
     const loading = isCommitting === true ? <Loading /> : undefined
-    const commitVerb = loading ? 'Committing' : 'Commit'
+
+    const isAmending = this.props.commitToAmend !== null
+    const commitVerb = isAmending
```
Comment (tidy-dev) [inline app/src/ui/changes/commit-message.tsx]: This one isn't as hard to follow but.. maybe still clearer if broken into a variable for each type like below.

File: app/src/ui/changes/sidebar.tsx [hunk 4]
```diff
@@ -299,7 +300,8 @@ export class ChangesSidebar extends React.Component<IChangesSidebarProps, {}> {
 
     // We don't allow undoing commits that have tags associated to them, since then
     // the commit won't be completely deleted because the tag will still point to it.
-    if (commit && commit.tags.length === 0) {
+    // Also, don't allow undoing commits while the user is amending the last one.
+    if (commit && commit.tags.length === 0 && !this.props.isAmending) {
       child = (
         <CSSTransition
```
Comment (tidy-dev) [inline app/src/ui/changes/sidebar.tsx:306]: TIL: We have CSSTransition thing!

**Review by tidy-dev — CHANGES_REQUESTED** [review]
This looks great and I pulled it down and tested it out. It is awesome!

I just a small suggestion on the double tertiary's. The first one not so much, but the second one feels really hard to process when reading it.

**Comment by sergiou87** [comment]
Addressed both issues, thank you! ❤️

**Review by tidy-dev — DISMISSED** [review]
✨  Thank you for those changes!

**Review by ampinsk — COMMENTED** [review]
Overall this looks great! We chatted in Slack about updating the text above the commit button to: 
> Your changes will modify your most recent commit. Stop amending to make these changes as a new commit.

File: app/src/ui/changes/commit-message.tsx [hunk 5]
```diff
@@ -569,19 +598,29 @@ export class CommitMessage extends React.Component<
...
 
-    if (showNoWriteAccess) {
+    if (commitToAmend !== null) {
       return (
-        <PermissionsCommitWarning>
+        <CommitWarning icon={CommitWarningIcon.Information}>
+          Your changes will be applied to your{' '}
+          <strong>most recent commit</strong>.{' '}
+          <LinkButton onClick={this.onStopAmending}>Stop amending</LinkButton>{' '}
+          to go back to normal.
```
Comment (billygriffin) [inline app/src/ui/changes/commit-message.tsx]: From @ampinsk's suggestion from workshopping in Slack.
```suggestion
          Your changes will modify your{' '}
          <strong>most recent commit</strong>.{' '}
          <LinkButton onClick={this.onStopAmending}>Stop amending</LinkButton>{' '}
          to make these changes as a new commit.
```

**Review by tidy-dev — APPROVED** [review]
:sparkles: This is so nice! Looking forward to using this feature.  I reviewed recent commits and pulled down and retested. LGTM!

**Comment by MichaIng** [comment]
> the ability to amend the last commit in your branch (if it's not pushed)

What is the reason for not allowing to amend to pushed commits? I regularly run into issues with GitHub desktop where it doesn't allow things for none obvious reasons which work fine with `git`. Sometimes we just want it to do what we want, and as long as it does not allow steps that can be done fine with `git`, it's very difficult for users to switch, running into a bunch of unexpected difficulties 🤔.

**Comment by sergiou87** [comment]
We didn't want to create this kind of "force-push" situations very often, but we've talked about it and we'll allow to amend pushed commits 😄 

Thank you for your feedback @MichaIng !!

**Comment by MichaIng** [comment]
Many thanks 👍!

At least I use this quite often to squash or repair commits in a PR without bloating the commit history. Currently I am forced to use `git` CLI for that, and sadly on an additional Linux system since `git` for Windows has a bunch of issues related to Windows 10 security/sandboxing features. Much easier to do everything from the same desktop system where I run notepad++ and browse GitHub 🙂.

**Comment by sergiou87** [comment]
@MichaIng hopefully will get merged soon: https://github.com/desktop/desktop/pull/13384 😄
